Use one tunnel per connection instead of multiplexing them into one tunnel.

thallgren opened the pull request:
## Description

This PR is a refactoring of the Telepresence tunneling framework. It's main objective is to increase performance by deprecating the old solution which multiplexed all connections over the same tunnel and instead use one tunnel per connection.

I found several very minor issues during the work to switch this over, and rather than creating one PR for each little thing I found, I decided to create a commit here even if some of the them aren't directly related to the tunnels.

I recommend reviewing one commit at a time and to use the "ignore white space" mode when reviewing.

## Checklist

<!--
  Please review the requirements for each checkbox, and check them
  off (change "[ ]" to "[x]") as you verify that they are complete.
-->

 - [ ] I made sure to update `./CHANGELOG.md`.
 - [ ] I made sure to either submit a docs PR, or tell Matt about the necessary documentation changes.
 - [ ] My change is adequately tested.
 - [ ] I updated `DEVELOPING.md` with any any special dev tricks I had to use to work on this code efficiently.
 - [ ] I updated `TELEMETRY.md` if I added, changed, or removed a metric name.

Commit:
Improve DNS lookup performance.

DNS lookups where serialized in the agent. A bad request therefore
blocked good requests for some time. Also, a lookup was performed in
the traffic-manager even though an agent had responded. An agent
response should be authoritative even if it didn't find anything.

This commit ensures that the agent produces lookup responses in a
separate goroutine to ensure that multiple lookups can run in parallel
and ensures that no lookup is made by the traffic-manager if responses
have arrived from at least one agent.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Rename pkg/tun to pkg/vif

VIF is the commonly used abbreviation for Virtual Network
Interface and a more accurate description of this package.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Rename connpool.Tunnel to connpool.MuxTunnel

In preparation for a tunnel package.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Move generic tunnel logic from connpool to new tunnel package

In preparation for deprecating the connpool package.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Minor cleanups and logging improvements in manager and agent.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Log when sessions are removed in traffic-manager.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Add LoadAllMatching(filter-func) to the whatchable maps.

The LoadAllMatching enables a concurrency-safe load of all entries
for which a filter function returns true. This can improve performance
when selecting few entries for larger maps since all returned entries
are deep cloned.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Make all sessionStates lockable, not just clientSessionState.

This is in preparation for code that will require that the
agentSessionState can also be locked.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Use RWMutex in the tunnel.Pool() and add error return to Handler.Start()

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Don't panic on bogus ConnID's String() method.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Use common MTU calculation (1500 - 40 == 1460 bytes for L4 payload)

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Make the mgrProxy a private immutable, and simplify its creation.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Fix better error handling for intercept and agent watchers.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Ensure that scout reporting cannot slow down DNS resolution

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Add offset to TUN device's WritePacket method.

The method returns the number of bytes of a packet that were written so
it makes sense to be able to write more based on an offset in case
there's more bytes in the packet.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Make reading of TUN packets more efficient.

Split the load between the TUN device read loop, the dispatcher, and
the handlers by smarter use of channels. Also ensure that write of
larger packages is handled correctly.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Change how traffic-manager finds agents intercepted by client.

The getInterceptedAgents() method must not be dependent on
the agentTunnels because they will be deprecated.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Add timeouts for roundtripLatency and endpointDial.

These timeouts are propagated in the intercept spec and used in future
commits when doing remote dials, both from the client and the
traffic-agent.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Rename "package" to "packet" where applicable.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Add a timestamping test logger.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Rename connForward to muxTunnelForward.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Add Tunnel and WatchDial to gRPC.

A Tunnel represents one single connection where the client or
traffic-agent represents one end (the client side) and the
traffic-manager represents the other (the server side). The first
message that a client sends when the tunnel is established is will
always contain the session ID, connection ID, and timeouts used by
the dialer endpoints.

`WatchDial` makes it possible for the client side to receive
`DialRequest`s from the `traffic-manager`. Requests are sent when an
intercepted `traffic-agent` creates a `Tunnel` that needs to be extended
to the Telepresence client on the workstation, or the other way around,
when that client creates a tunnel that needs to be extended to an
intercepted `traffic agent`. The receiver of the request dials a
connection and responds with a `Tunnel`. The manager then connects the
two tunnels.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Add connection based Stream implementation

Adds the Stream implementation that is used on top of by the
manager.Tunnel and it's `BidiPipe` and `Dialer` endpoints.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Add DialRequest watchers to the traffic-agent and the connector daemon.

This also means that from now on, it's the connector and not the daemon
that performs the dials. Which makes sense, since the dials should be
performed by the user instead of root.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Add traffic-manager's implementation of Tunnel and WatchDial.

Adds the Tunnel and WatchDial implementations to the `traffic-manager`
so that it is capable of creating and connecting tunnel's, and adding
dialer endpoints to the tunnels.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Add implementation of Tunnel and WatchDial to the manager proxy.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Enable version detection using mux-tunnel.

Enables the guessing of version from mux-tunnel implementations older
than 2.4.4 using a push-back mechanism for the first message that is
received.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Choose tunnel implementation depending on peer version.

Adds the ability to use Stream instead of a MuxTunnel in the tcp and
udp handlers and makes the traffic-manager choose the best one based
on the peer's tunnel version.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Use trace-level logging for normal network processing.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Mark the MuxTunnel implementation as deprecated.

The MuxTunnel implementation will be removed when we drop support
for Telepresence <= 2.4.4 so all its components are marked deprecated
by this commit.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Move tcp's handler.writeToTunLoop to tmgr.readFromMgrMux()

No code changes, just a move to get it adjacent to its corresponding
Stream based function `readFromMgrLoop()`.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Fix bugs in the TCP-sliding window implementation.

Separates our window scale from the peer's window scale and ensures that
the window size is updated and propagated correctly.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Loose packages instead of blocking and keep track of them.

If the manager doesn't keep up, then we cannot block the TUN-device
by first filling all channels and then stop reading because that would
block all connections, not just the one that isn't keeping up. So
when that happens (in spite of having the receive window set to zero)
we actually loose packages and expect the sender to resend them since
they are not ACKed.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Defer setting sequence and ack until the TCP package is actually sent.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Fix some glitches regarind TCP RST and what happens during timed wait.

An RST is quitByReset, not quitByUs.
A timedWait must process incoming normally.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Add better control over how the TCP-connection ends.

Adds a new `tunDone` channel which closes when the TUN-device is closed
due to packet negotiation. Also adds a WaitGroup to ensure that the
connection remains alive until all loops have ended.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Never hang TCP packet handling on a pending write to the manager.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Increase total time to run `go test` (and the ci test)

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

donnyyung commented on pkg/tunnel/pool.go:
```diff
@@ -10,56 +10,59 @@ import (
 	Close(context.Context)
 
-	Start(ctx context.Context)
+	Start(ctx context.Context) error
```
Might just be where I'm at in the commit-by-commit review (abd2681c328f745f22df107eaaf791152b6e0b64), but you added error as an output and I don't currently see any `errors` returned in the start functions edited in this commit... so not sure of the purpose.

thallgren replied:
You're right. There's no need for an error return.

josecv commented on cmd/traffic/cmd/agent/client.go:
```diff
@@ -157,7 +159,7 @@ func lookupHostWaitLoop(ctx context.Context, manager rpc.ManagerClient, session
 		if err != nil {
-			if ctx.Err() == nil {
+			if ctx.Err() == nil && !errors.Is(err, io.EOF) {
 				dlog.Debugf(ctx, "lookup request stream recv: %+v", err) // May be io.EOF
```
we should probably update the `May be io.EOF` comment if it won't be io.EOF

josecv commented on pkg/tunnel/pool.go:
```diff
@@ -69,25 +72,29 @@ func (p *Pool) GetOrCreate(ctx context.Context, id ConnID, createHandler Handler
+	if err = handler.Start(handlerCtx); err != nil {
+		release()
+		return nil, false, err
+	}
```
should we remove it from `p.handlers` if it fails to start?

thallgren replied:
We should, and that's exactly what the call to `release()` does.

thallgren replied:
Moot point now, since `Start()` no longer returns error.

josecv commented on pkg/client/daemon/outbound.go:
```diff
@@ -48,7 +48,7 @@ type outbound struct {
 	dnsConfig *rpc.DNSConfig
 
-	scout chan scout.ScoutReport
+	scout chan<- scout.ScoutReport
```
TIL about `<-chan` and `chan<-`

josecv commented on pkg/client/daemon/tunrouter.go:
```diff
@@ -452,7 +447,30 @@ func (t *tunRouter) handlePacket(c context.Context, data *buffer.Data) {
+	*vif.Device
+}
+
+func (w vifWriter) Write(ctx context.Context, pkt ip.Packet) (err error) {
```
does this need to softRelease the packet?

thallgren replied:
No, it's supposedly taken care of by the caller. But thanks for the reminder. The `Packet.SoftRelease()` should be removed completely since it's no longer needed.

josecv commented on pkg/log/testlogger.go:
```diff
@@ -0,0 +1,66 @@
+}
+
+func (w *tbWrapper) WithField(key string, value interface{}) dlog.Logger {
+	ret := tbWrapper{
+		TB:     w.TB,
+		fields: make(map[string]interface{}, len(w.fields)+1),
+	}
+	for k, v := range w.fields {
+		ret.fields[k] = v
+	}
+	ret.fields[key] = value
+	return &ret
+}
+
+func (w *tbWrapper) Log(level dlog.LogLevel, msg string) {
+	if level > w.level {
+		return
+	}
+	w.Helper()
+	sb := strings.Builder{}
+	sb.WriteString(time.Now().Format("15:04:05.0000"))
+	sb.WriteString(" ")
+	sb.WriteString(msg)
+
+	if len(w.fields) > 0 {
+		parts := make([]string, 0, len(w.fields))
+		for k := range w.fields {
+			parts = append(parts, k)
+		}
+		sort.Strings(parts)
+
+		for i, k := range parts {
+			if i > 0 {
+				sb.WriteString(" ")
+			}
+			fmt.Fprintf(&sb, "%s=%#v", k, w.fields[k])
+		}
+	}
+	w.TB.Log(sb.String())
+}
```
i don't really have an issue with this, but the weird thing is that the dlib test logger should add timestamps https://github.com/datawire/dlib/blob/master/dlog/logger_testing.go#L71-L73 -- maybe we should fix it there? there seems to be some bug where some log lines get timestamped and others do not

thallgren replied:
The dlib testlogger uses fields and escaped quoted strings (lots of `\"`) which makes the timestamps unaligned and the lines hard to read even if we fix the lack-of-timestamp problem. The `log/testlogger.go` is backported from the integration test refactoring that I'm working on and what I'm after is something more similar to other logs that Telepresence outputs. I'm all for suggesting that the dlog test-logger would do the same, but I also think it might be good to show what it would look like.

josecv reviewed (commented):
so im not yet done reviewing this, but i wanna post my comments so far (i'm up to `aeb2c353a6574f03b853e1681468fd181663c11e`) so that you can take a look at em since i  won't be here on monday

Commit:
Don't log EOF in agent wait loops and drop comment about doing so.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Get rid of Packet.SoftRelease()

The tcp packet handler now handles the ack-queue explicitly when sending
packages to the VIF and there's no longer any need for a specific "soft"
release of packages.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Drop error return from Handler.Start(). It's never used.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Don't redirect log output to Tempdir() in telepresence_test

There's really no reason to use a special directory for the logfiles
just because the test checks what it contains. A new logfile is created
on each run anyway. The logfiles are normally gathered at the end of the
run and using a temporary directory prevents that. Furthermore, there's
a fair chance that a daemon process isn't quite dead yet when the tests
end which often yields weird cleanup errors when deleting the tempdir
(especially on Windows).

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Remove ambiguity when using streams returned by tunnel.ReadLoop()

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Increase the CI "wait for cluster" time to 30 minutes.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Make StreamInfo and StreamOK package private.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Traffic-manager should not pretend that missing session is cancelled.

If the traffic-manager is restarted and old clients and agents try
to connect to it, it will not find the sessions. That's different from
having an established session that gets cancelled. This commit ensures
that the proper NotFound error is returned when the session doesn't
exist.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Fix incorrect error handling in agent's interceptWaitLoop()

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Simplify condition for logging error in agentInfoWatcher

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Fix broken test and lint error.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

donnyyung commented:
Mostly some tiny nits, want to do some manual testing tomorrow and may do a few more sanity checks, there's still some gaps in my knowledge (around networking concepts, not anything specific in your PR) that I'd like to fill in.

Commit:
Log version of already installed traffic-manager during connect.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Fix backward compatibility problem with traffic-managers <= 2.4.2.

Deduct tunnel version from the version reported by the traffic-manager
instead of trying to guess based on tunnel packages.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Fix test that panics due to previous asserted error.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Add missing defer for `wg.Done()` call.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

josecv commented on pkg/client/daemon/tunrouter.go:
```diff
@@ -507,11 +527,35 @@ func (t *tunRouter) udp(c context.Context, dg udp.Datagram) {
+
+func (t *tunRouter) maybeOpenStream(c context.Context, id tunnel.ConnID) (tunnel.Stream, error) {
+	if t.muxTunnel != nil {
+		// tunnelVersion <= 2, so use the multiplexing tunnel
```
should this  be `< 2`?

josecv commented on pkg/vif/tcp/handler.go:
```diff
@@ -342,31 +331,28 @@ func (h *handler) processPayload(ctx context.Context, data []byte) {
-			if window == 0 {
+		window := int(h.peerWindow) - int(h.sequence()-h.seqAcked)
+		for window <= 0 {
+			if window <= 0 {
```
do we need the if? there's no more `||` in the loop invariant so it seems we could just directly log

josecv reviewed (approved):
woo i think this lgtm now

i really like the new tunnel semantics, i feel like the interfaces are a whole lot cleaner than the previous package -- what are your thoughts around getting rid of the deprecated muxTunnel? It seems there's a whole lot of places that check for it, so i think it might be good to have a plan for eventually getting rid of it, esp if we need to be adding new features to it

thallgren commented:
@josecv I *really want* to get rid of the deprecated muxTunnel a.s.a.p. But we must provide a window of backward compatibility for some reasonable time. The fact that an explicit Helm install has precedence, combined with users that act on the "there's a new version available" message, will probably make it common to have a 2.4.5 talking to an older manager.

I think we should inform the user about the deprecation and when we plan to end support for it for some, yet to be determined, period of time, before we remove it.

Commit:
Changes in response to code-review.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>
